perfschema/perfschema.go

```diff
@@ -33,40 +31,69 @@ type StatementInstrument interface {
 
 // PerfSchema defines the methods to be invoked by the executor
 type PerfSchema interface {
-	// For SELECT statement only.
-	NewPerfSchemaPlan(tableName string) (plan.Plan, error)
-
-	// For INSERT statement only.
-	ExecInsert(insertVals *InsertValues) error
 
 	// For statement instrumentation only.
 	StatementInstrument
+
+	// GetDBMeta returns db info for PerformanceSchema.
+	GetDBMeta() *model.DBInfo
+	// GetTable returns table instance for name.
+	GetTable(name string) (table.Table, bool)
 }
 
 type perfSchema struct {
-	tables map[string]*model.TableInfo
-	fields map[string][]*field.ResultField
-	stores map[string]*leveldb.DB
-	lsns   map[string]*uint64
+	store   kv.Storage
+	dbInfo  *model.DBInfo
+	tables  map[string]*model.TableInfo
+	mTables map[string]table.Table // MemoryTables for perfSchema
+
+	// Used for TableStmtsHistory
+	historyHandles []int64
+	historyCursor  int
 }
 
 var _ PerfSchema = (*perfSchema)(nil)
 
+// PerfHandle is the only access point for the in-memory performance schema information
 var (
-	pool = sync.Pool{
-		New: func() interface{} {
-			return &leveldb.Batch{}
-		},
-	}
+	PerfHandle PerfSchema
 )
 
-// PerfHandle is the only access point for the in-memory performance schema information
-var PerfHandle PerfSchema
+// NewPerfHandle creates a new perfSchema on store.
+func NewPerfHandle(store kv.Storage) PerfSchema {
+	schema := PerfHandle.(*perfSchema)
+	schema.store = store
+	schema.historyHandles = make([]int64, 0, stmtsHistoryElemMax)
+	_ = schema.initialize()
+	// Existing instrument names are the same as MySQL 5.7
+	PerfHandle.RegisterStatement("sql", "alter_table", (*ast.AlterTableStmt)(nil))
```

Comment:
I would prefer move this `RegisterStatement` calls to `statement.go` since it's all related to statement instruments.